modules/programs/thunderbird.nix

```diff
@@ -0,0 +1,269 @@
+              mkEnableOption "the Thunderbird mail client for this account";
+
+            profiles = mkOption {
+              type = with types; listOf str;
+              default = [ ];
+              example = literalExpression ''
+                [ "profile1" "profile2" ]
+              '';
+              description = ''
+                List of Thunderbird profiles for which this account should be
+                enabled. If this list is empty (the default), this account will
+                be enabled for all declared profiles.
+              '';
+            };
+
+            settings = mkOption {
+              type = with types; functionTo (attrsOf (oneOf [ bool int str ]));
+              default = _: { };
+              defaultText = literalExpression "_: { }";
+              example = literalExpression ''
+                id: {
+                  "mail.identity.id''${id}.protectSubject" = false;
+                  "mail.identity.id''${id}.autoEncryptDrafts" = false;
+                };
+              '';
+              description = ''
+                Extra settings to add to this Thunderbird account configuration.
+                The <varname>id</varname> given as argument is an automatically
+                generated account identifier.
+              '';
+            };
+          };
+        });
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      (hm.assertions.assertPlatform "programs.thunderbird" pkgs platforms.linux)
+
+      (let defaults = catAttrs "name" (filter (a: a.isDefault) profilesWithId);
+      in {
+        assertion = cfg.profiles == { } || length defaults == 1;
+        message = "Must have exactly one default Thunderbird profile but found "
+          + toString (length defaults) + optionalString (length defaults > 1)
+          (", namely " concatStringsSep "," defaults);
+      })
+
+      (let
+        profiles = catAttrs "name" profilesWithId;
+        selectedProfiles =
+          concatMap (a: a.thunderbird.profiles) enabledAccounts;
+      in {
+        assertion = (intersectLists profiles selectedProfiles)
+          == selectedProfiles;
+        message = "Cannot enable an account for a non-declared profile. "
+          + "The declared profiles are " + (concatStringsSep "," profiles)
+          + ", but the used profiles are "
+          + (concatStringsSep "," selectedProfiles);
+      })
+    ];
+
+    home.packages = [ cfg.package ]
+      ++ optional cfg.withExternalGnupg pkgs.gpgme;
+
+    home.file = mkMerge ([{
+      "${thunderbirdConfigPath}/profiles.ini" =
+        mkIf (cfg.profiles != { }) { text = generators.toINI { } profilesIni; };
+    }] ++ flip mapAttrsToList cfg.profiles (name: profile: {
+      "${thunderbirdConfigPath}/${name}/user.js" = let
+        accounts = filter (a:
+          a.thunderbird.profiles == [ ]
+          || any (p: p == name) a.thunderbird.profiles) enabledAccountsWithId;
+      in mkIf (cfg.settings != { } || length accounts != 0) {
```

Comment:
just having profile settings should be sufficient here? Also it's maybe a bit silly to create profile in the profiles.ini and nothing else?

Reply:
> just having profile settings

You mean only profile-specific settings and no general `settings` option? If so, I just thought it was cleaner than having users handle "general" and "per-profile" settings by themselves. Do you think I should remove this option?

> create profile in the profiles.ini and nothing else?

Can the `profiles.ini` file declare something else? If so, what?

Reply:
I mean without general settings no user.js is being generated with this condition so profile just adds itself to profiles.ini and "nothing else", I don't think profiles.ini needs anything else.

Reply:
Oh, that's right. I guess I can just remove the condition, and a profile without enabled accounts or configuration would just have an empty `user.js` file.

Reply:
I guess someone might have just manually configured profile or something, not sure how that works, not sure if corner cases like this are subject to backwards compat (maybe even assert and make the person hitting the assertion explain the use case).

I'm fine with any solution that does not require the general settings and can manage with just profile settings.

Reply:
Also extending profiles.ini is also possible later I guess.